sd/lb/retry.go

```diff
@@ -10,12 +10,32 @@ import (
 	"github.com/go-kit/kit/endpoint"
 )
 
+// Callback function that indicates the current attempt count and the error encountered.
+//  Should return whether the Retry function should continue trying, and a custom
+//  error message if desired. The error message may be nil, but a true/false
+//  is always expected. In all cases if the error message is supplied, the
+//  current error will be replaced.
+type Callback func(n int, err error) (cont bool, cbErr error)
+
 // Retry wraps a service load balancer and returns an endpoint oriented load
 // balancer for the specified service method.
 // Requests to the endpoint will be automatically load balanced via the load
 // balancer. Requests that return errors will be retried until they succeed,
 // up to max times, or until the timeout is elapsed, whichever comes first.
 func Retry(max int, timeout time.Duration, b Balancer) endpoint.Endpoint {
+	return RetryWithCallback(max, timeout, b, func(c int, err error) (bool, error) { return true, nil })
+}
```

Comment:
I've split the max retry logic out into the callback. There's a helper function for generating a max retries callback if that's all you want to do, and `Retry()` passes one of those on to `RetryWithCallback()` now.

I'm not sure if or how to tackle handling timeouts in the callback. I think it's reasonable to keep those two jobs separate. To me, I'm happy with the timeout in the `select` inside `RetryWithCallback()`. What do you reckon?